mesonbuild/modules/qt.py

```diff
@@ -170,12 +240,19 @@ def parse_qrc_deps(self, state, rcc_file):
                     result.append(File(is_built=False, subdir=state.subdir, fname=path_from_rcc))
         return result
 
-    @noPosargs
-    @permittedKwargs({'method', 'required'})
     @FeatureNew('qt.has_tools', '0.54.0')
-    def has_tools(self, state, args, kwargs):
+    @noPosargs
+    @typed_kwargs(
+        'qt.has_tools',
+        KwargInfo('required', (bool, FeatureOptionHolder), default=False),
+        KwargInfo('method', str, default='auto'),
+    )
+    def has_tools(self, state: 'ModuleState', args: T.Tuple, kwargs: 'HasToolKwArgs') -> bool:
         method = kwargs.get('method', 'auto')
-        disabled, required, feature = extract_required_kwarg(kwargs, state.subproject, default=False)
+        # We have to cast here because TypedDicts are invariant, even though
+        # ExtractRequiredKwArgs is a subset of HasToolKwArgs, type checkers
+        # will insist this is wrong
+        disabled, required, feature = extract_required_kwarg(T.cast('ExtractRequiredKwArgs', kwargs), state.subproject, default=False)
```

Comment:
Haven't tested it, but can't you fix this with something like this:

```py
class HasToolKwArgs(ExtractRequiredKwArgs):
    method: str
```

Reply:
I'm not sure, possibly? This is a case I might have to open an issue with upstream, because it seems like that should be valid, though the reality is I find python's `TypedDict` to fall far short of TypeScript's object annotations in general, where this would Just Work™

Reply:
OK, I have tested it and we can get rid of the `T.cast` if `HasToolKwArgs` inherits from `ExtractRequiredKwArgs`

Reply:
yay, that's nice.